tests/vptranslator.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright 2017 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+use chrono::{Duration, Weekday};
+use std::collections::BTreeSet;
+use transit_model;
+use transit_model::objects::{Date, ExceptionType, ValidityPeriod};
+use transit_model::vptranslator::{translate, ExceptionDate};
+
+//       July 2012
+// Mo Tu We Th Fr Sa Su
+//                    1
+//  2  3  4  5  6  7  8
+//  9 10 11 12 13 14 15
+// 16 17 18 19 20 21 22
+// 23 24 25 26 27 28 29
+// 30 31
+
+fn get_dates_from_bitset(start_date: Date, vpattern: &str) -> BTreeSet<Date> {
+    let mut res = BTreeSet::new();
+    for (i, c) in vpattern.chars().enumerate() {
+        if c == '1' {
+            res.insert(start_date + Duration::days(i as i64));
+        }
+    }
+    res
+}
+
+fn get_week_from_weekday(weekday: Vec<Weekday>) -> u8 {
+    let mut res = 0;
+    for day in weekday {
+        res |= 1 << (7 - day.number_from_monday());
+    }
+    res
+}
+
+#[test]
+fn only_first_day() {
+    let mut dates = BTreeSet::new();
+
+    dates.insert(Date::from_ymd(2012, 7, 2));
+    let res = translate(&dates);
+    assert_eq!(get_week_from_weekday(res.operating_days), 0b1000000);
+
+    assert!(res.exceptions.is_empty());
+    assert_eq!(
+        res.validity_periods.iter().next().unwrap(),
+        &ValidityPeriod {
+            start_date: Date::from_ymd(2012, 7, 2),
+            end_date: Date::from_ymd(2012, 7, 2),
+        }
+    );
+}
+
+#[test]
+fn bound_cut() {
+    let res = translate(&get_dates_from_bitset(
+        Date::from_ymd(2012, 7, 16),
+        &format!("{}{}", "0011101", "000"),
```

Comment:
why do you need this `format!` since all the values are hardcoded ?

Reply:
Reading the test is more visible because you can see the weeks